kube-client/src/config/incluster_config.rs

```diff
@@ -29,6 +37,43 @@ pub fn kube_dns() -> http::Uri {
     http::Uri::from_static("https://kubernetes.default.svc/")
 }
 
+pub fn try_kube_from_env() -> Result<http::Uri, Error> {
+    // client-go requires that both environment variables are set.
+    let host = std::env::var("KUBERNETES_SERVICE_HOST").map_err(Error::ReadEnvironmentVariable)?;
+    let port = std::env::var("KUBERNETES_SERVICE_PORT")
+        .map_err(Error::ReadEnvironmentVariable)?
+        .parse::<u16>()
+        .map_err(Error::ParseClusterPort)?;
+
+    // Format a host and, if not using 443, a port.
+    //
+    // Ensure that IPv6 addresses are properly bracketed.
+    const HTTPS: &str = "https";
+    let uri = match host.parse::<std::net::IpAddr>() {
+        Ok(ip) => {
+            if port == 443 {
+                if ip.is_ipv6() {
+                    format!("{HTTPS}://[{ip}]")
+                } else {
+                    format!("{HTTPS}://{ip}")
+                }
+            } else {
+                let addr = std::net::SocketAddr::new(ip, port);
+                format!("{HTTPS}://{addr}")
+            }
+        }
+        Err(_) => {
+            if port == 443 {
+                format!("{HTTPS}://{host}")
+            } else {
+                format!("{HTTPS}://{host}:{port}")
+            }
+        }
+    };
+
+    uri.parse().map_err(Error::ParseClusterUrl)
```

Comment:
We had some tests for this before in https://github.com/kube-rs/kube-rs/commit/dd0b2585729dab5c140ab96dc35c00484cc992bc#diff-305bd48dd4adbbd7c75ef13d11d4fa16ad1015378eb9cef735e757475554a0bbL93-L128

I think having those tests lock down the behaviour is probably valuable still.

Reply:
I wrote some tests but I'm noticing that mutating `std::env` can be racey/flakey, since tests may run in parallel and do not run in separate processes... So, to test this, we need to refactor the helper a bit.

Reply:
I see you've put the testable part in its own function, so that should be fine for now!